File: lib/host/clipfish.c
/* 
* Copyright (C) 2009,2010 Toni Gundogdu.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "host.h"

_host_constants(clipfish, "clipfish.de", "flv");

_host_re(re_id,     "(?i)\\/video\\/(.*?)\\/");
_host_re(re_title,  "(?i)<title>(.*?)</title>");
_host_re(re_lnk,    "(?i)&url=(.*?)&");

QUVIcode
handle_clipfish(const char *url, _quvi_video_t video) {
    char *content, *config_url, *config, *lnk;
    QUVIcode rc;

    /* host id */
    _host_id("clipfish");

    /* common */
    rc = parse_page_common(url, video, &content, re_id, re_title);
    _free(content);

    if (rc != QUVI_OK)
        return (rc);

    video->title = strepl(video->title, "Video: ", "");
    video->title = strepl(video->title, " - Clipfish", "");

    /* config */
    asprintf(&config_url,
        "http://www.clipfish.de/video_n.php?p=0|DE&vid=%s", video->id);

    rc = fetch_to_mem(video, config_url, QUVISTATUSTYPE_CONFIG, &config);
    _free(config_url);

    /* video link */
    rc = regexp_capture(
        video->quvi,
        config,
        re_lnk,
        0,
        0,
        &lnk,
        (void *) 0
    );

    _free(config);

    if (rc != QUVI_OK)
        return (rc);

    rc = add_video_link(&video->link, "%s", lnk);

    _free(lnk);

    return (rc);
}

File: share/lua/website/clipfish.lua

-- quvi
-- Copyright (C) 2010  Toni Gundogdu <legatvs@gmail.com>
--
-- This file is part of quvi <http://quvi.sourceforge.net/>.
--
-- This library is free software; you can redistribute it and/or
-- modify it under the terms of the GNU Lesser General Public
-- License as published by the Free Software Foundation; either
-- version 2.1 of the License, or (at your option) any later version.
--
-- This library is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-- Lesser General Public License for more details.
--
-- You should have received a copy of the GNU Lesser General Public
-- License along with this library; if not, write to the Free Software
-- Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
-- 02110-1301  USA
--

-- Identify the script.
function ident (self)
    package.path = self.script_dir .. '/?.lua'
    local C      = require 'quvi/const'
    local r      = {}
    r.domain     = "clipfish.de"
    r.formats    = "default"
    r.categories = C.proto_http
    local U      = require 'quvi/util'
    r.handles    = U.handles(self.page_url, {r.domain}, {"/video/%d+/"})
    return r
end

-- Parse video URL.
function parse (self)
    self.host_id = "clipfish"
    local page   = quvi.fetch(self.page_url)

    local _,_,s = page:find("<title>(.-)</title>")
    self.title  = s or error ("no match: video title")

    self.title = self.title:gsub("Video:%s+", "")
    self.title = self.title:gsub("%s+-%s+Clipfish", "")

    local _,_,s = page:find("/video/(.-)/")
    self.id     = s or error ("no match: video id")

    local config_url =
        string.format("http://www.clipfish.de/video_n.php?p=0|DE&vid=%s",
            self.id)

    local config = quvi.fetch (config_url, {fetch_type = 'config'})
    local _,_,s  = config:find("&url=(.-)&")
    self.url     = {s or error ("no match: url")}

    return self
end

-- vim: set ts=4 sw=4 tw=72 expandtab:
